### PR TITLE
feat: optional html quotes escaping for HTMLRenderer (#176)

### PR DESCRIPTION
**Pull request by pbodnar**
Introduce additional parameters for the `HTMLRenderer` constructor that control double and single quotes escaping.
Also revert to previous behavior when single quotes are not escaped (by default).

### CODE REVIEW

File: mistletoe/html_renderer.py [hunk 1]
```diff
@@ -193,12 +199,20 @@ def render_document(self, token: block_token.Document) -> str:
...
         """
-        This method is deprecated. Use `html.escape` instead.
+        Like `html.escape()`, but this  looks into the current rendering options
+        to decide which of the quotes (double, single, or both) to escape.
+
+        Intended for escaping text content. To escape content of an attribute,
+        simply call `html.escape()`.
         """
-        return html.escape(raw)
+        s = html.escape(s, quote=False)
```
Comment (pbodnar) [inline mistletoe/html_renderer.py:210]: As `html.escape` doesn't seem to be a native function, in-lining it here should be faster?

File: mistletoe/contrib/mathjax.py [hunk 2]
```diff
@@ -6,10 +6,17 @@
 from mistletoe.latex_renderer import LaTeXRenderer
 
 class MathJaxRenderer(HTMLRenderer, LaTeXRenderer):
-    """
-    MRO will first look for render functions under HTMLRenderer,
-    then LaTeXRenderer.
-    """
+    def __init__(self, **kwargs):
+        """
+        MRO will first look for render functions under HTMLRenderer,
```
Comment (anderskaplan) [inline mistletoe/contrib/mathjax.py:11]: Maybe "MRO" deserves an explanation?

File: mistletoe/html_renderer.py [hunk 3]
```diff
@@ -18,13 +18,19 @@ class HTMLRenderer(BaseRenderer):
 
     See mistletoe.base_renderer module for more info.
     """
-    def __init__(self, *extras):
+    def __init__(self, *extras, html_escape_double_quotes=True, html_escape_single_quotes=False):
```
Comment (anderskaplan) [inline mistletoe/html_renderer.py:21]: "html-escape" or just "escape"? Imho, escape is sufficient.

File: mistletoe/html_renderer.py [hunk 4]
```diff
@@ -193,12 +199,20 @@ def render_document(self, token: block_token.Document) -> str:
...
+
+        Intended for escaping text content. To escape content of an attribute,
+        simply call `html.escape()`.
         """
-        return html.escape(raw)
+        s = html.escape(s, quote=False)
+        if self.html_escape_double_quotes:
+            s = s.replace('"', "&quot;")
+        if self.html_escape_single_quotes:
+            s = s.replace('\'', "&#x27;")
```
Comment (anderskaplan) [inline mistletoe/html_renderer.py:214]: Maybe `&apos;` instead of the ascii value, similar to the double quote?

File: mistletoe/html_renderer.py (hunk 1 above)
Comment (anderskaplan) [reply mistletoe/html_renderer.py:210]: An advantage of inlining it would be to avoid the clever but scary patching of `html._charref` done in `span_tokenizer.tokenize()`.

File: mistletoe/html_renderer.py (hunk 4 above)
Comment (anderskaplan) [reply mistletoe/html_renderer.py:214]: Ah now I see this is how it's done in `html.escape()`. So there's probably a good reason for it. Never mind.

File: mistletoe/html_renderer.py (hunk 3 above)
Comment (anderskaplan) [inline mistletoe/html_renderer.py:21]: Should the `HTMLRenderer.init()` perhaps also take **kwargs parameters? And `Scheme`, too?

File: test/specification/commonmark.py [hunk 5]
```diff
@@ -34,7 +34,7 @@ def run_tests(test_entries, start=None, end=None,
 def run_test(test_entry, quiet=False):
     test_case = test_entry['markdown'].splitlines(keepends=True)
     try:
-        output = markdown(test_case).replace('&#x27;', "'")
```
Comment (anderskaplan) [inline test/specification/commonmark.py:37]: Does this mean that the CommonMark tests now pass without special handling of escaped quotes? If yes, then that's a big win!

File: test/test_html_renderer.py [hunk 6]
```diff
@@ -138,6 +138,19 @@ def test_document(self):
         self._test_token('Document', '', footnotes={})
 
 
+class TestHTMLRendererEscaping(TestCase):
+    @parameterized.expand([
```
Comment (anderskaplan) [inline test/test_html_renderer.py:142]: Nice way to put it.

File: mistletoe/html_renderer.py (hunk 4 above)
Comment (anderskaplan) [reply mistletoe/html_renderer.py:214]: Turns out [`&apos;` is an XML entity reference but not HTML](https://fishbowl.pastiche.org/2003/07/01/the_curse_of_apos). Odd.

File: mistletoe/contrib/mathjax.py (hunk 2 above)
Comment (pbodnar) [reply mistletoe/contrib/mathjax.py:11]: Yeah, maybe, it is <https://docs.python.org/3/library/stdtypes.html#class.__mro__>. But I didn't want to dig any deeper for now. :P

File: mistletoe/html_renderer.py (hunk 3 above)
Comment (pbodnar) [reply mistletoe/html_renderer.py:21]: I was asking myself the same question. :) And I decided to go with the longer variant, so I can be sure this won't collide for example with `latex_escape_double_quotes` or similar in the future, no matter how improbable this is now. ;)

File: mistletoe/html_renderer.py (hunk 1 above)
Comment (pbodnar) [reply mistletoe/html_renderer.py:210]: `html._charref` is used for `html.unescape` though? :)

File: test/specification/commonmark.py (hunk 5 above)
Comment (pbodnar) [reply test/specification/commonmark.py:37]: Yes. I would say a little win here. ;)

File: test/test_html_renderer.py (hunk 6 above)
Comment (pbodnar) [reply test/test_html_renderer.py:142]: Thanks. :)

**Review by pbodnar — COMMENTED** [review]
@anderskaplan, thanks for the review! 👍

File: mistletoe/html_renderer.py (hunk 1 above)
Comment (anderskaplan) [reply mistletoe/html_renderer.py:212]: Yes, you're right. It won't help for that.

File: mistletoe/contrib/mathjax.py (hunk 2 above)
Comment (anderskaplan) [reply mistletoe/contrib/mathjax.py:11]: How about changing "MRO" to "Python", then? It would make the comment understandable without digging into the technicalities.

Comment (anderskaplan) [reply mistletoe/contrib/mathjax.py:11]: (It is off-topic for this PR, for sure. But still could make sense, as the line is touched.)

Comment (pbodnar) [reply mistletoe/contrib/mathjax.py:11]: You;re right, we are moving it anyway. It is rather an inner workings note, explaining what Python does when there are multiple ancestors, so I would remove it altogether, what do you say?

Comment (anderskaplan) [reply mistletoe/contrib/mathjax.py:11]: Yes, sounds good!

Comment (pbodnar) [reply mistletoe/contrib/mathjax.py]: Done. :)

File: mistletoe/html_renderer.py (hunk 3 above)
Comment (pbodnar) [reply mistletoe/html_renderer.py]: Done.

File: mistletoe/html_renderer.py (hunk 1 above)
Comment (pbodnar) [reply mistletoe/html_renderer.py]: FYI, I in-lined the code in the end. Not that it would be visibly faster, but I guess it's the way to go in this case.

File: mistletoe/html_renderer.py (hunk 3 above)
Comment (pbodnar) [reply mistletoe/html_renderer.py]: Oddly enough, I can see now that I have somehow overlooked to also extend the constructor of `PygmentsRenderer` - will do additionally...
